### PR TITLE
[AIRFLOW-802] Add spark-submit operator/hook

### PR DESCRIPTION
**Pull request by Fokko**
Dear Airflow Maintainers,

I would like to add the spark-submit operator and hook. Right now we only support spark-sql operations by the SparkSqlOperator. Since spark-submit is a different binary, I've created a new hook and operator.
Besides the fact that spark-submit has much more options to configure, there might be features in the future that share functionality with the spark-sql operator.
One main implementation difference between the spark-sql and the spark-submit operator is the use of the connection_id of Airflow. It accepts a connection to set the master and the queue.

Please accept this PR that addresses the following issues:
https://issues.apache.org/jira/browse/AIRFLOW-802

Per Apache guidelines you need to create a [Jira issue](https://issues.apache.org/jira/browse/AIRFLOW/).

Testing Done:
- Unittests are required, if you do not include new unit tests please
specify why you think this is not required. We like to improve our
coverage so a non existing test is even a better reason to include one.

Reminders for contributors (REQUIRED!):
* Your PR's title must reference an issue on 
[Airflow's JIRA](https://issues.apache.org/jira/browse/AIRFLOW/). 
For example, a PR called "[AIRFLOW-1] My Amazing PR" would close JIRA 
issue #1. Please open a new issue if required!

* For all PRs with UI changes, you must provide screenshots. If the UI changes are not obvious, either annotate the images or provide before/after screenshots.

* Please squash your commits when possible and follow the [How to write a good git commit message](http://chris.beams.io/posts/git-commit/). 
Summarized as follows:
  1. Separate subject from body with a blank line
  2. Limit the subject line to 50 characters
  3. Do not end the subject line with a period
  4. Use the imperative mood in the subject line (add, not adding)
  5. Wrap the body at 72 characters
  6. Use the body to explain what and why vs. how

### CODE REVIEW

File: airflow/contrib/hooks/spark_submit_hook.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            connection_cmd += ["--queue", self._yarn_queue]
+
+        # The actual script to execute
+        connection_cmd += [application]
+
+        logging.debug("Spark-Submit cmd: {}".format(connection_cmd))
+
+        return connection_cmd
+
+    def submit(self, application="", **kwargs):
```
Comment (jhsenjaliya) [inline airflow/contrib/hooks/spark_submit_hook.py:168]: can this be implemented via bash_operator instead of its own subprocess ?

**Comment by jhsenjaliya** [comment]
Thanks for this PR !
do you think this could have been implemented by bash_operator or ssh_operator ? since its basically running a bash command locally or remotely ? 
since I wonder how we will keep track of spark command changes.

**Comment by Fokko** [comment]
Hi @jhsenjaliya 

You could of course implement this using a bash command. But in a production environment the spark-submit process will be light weight as the actual Spark driver is running on the YARN cluster inside of a container (when using `yarn-cluster`, which is most common). So it would not make sense to remotely execute the process on another machine.

The spark-submit command has been stable for a while. Many of the fine grained configuration options are submitted using the conf dict, if you like.

Does this answer your question, or am I missing something?

Cheers

**Comment by jhsenjaliya** [comment]
for remote execution one can also use ssh_operator.
I think once the spark-submit command is formed, its a matter of executing it either locally via bash or remotely via ssh operator. although building spark command based on parameters is useful functionality to get with this wrapper, it can execute with existing operator so that its easy to maintain afterwards due to reused code.
anyway, thats my 2 cents, I am not committer, so let see what others has to say in this. Thanks for explaining though.

**Comment by Fokko** [comment]
I value your input. This is some first contribution to Airflow beside a bugfix. So it could be that I'm missing something. How would you change the implementation?

I mostly based it on the other hooks which are already in there. For example, the sqoop and piq hooks are also using the subprocesses.

For maintainability I am thinking to merge the spark-sql and spark-submit by putting the shared configuration into an abstract class. But that's something for another pr ;-)

**Comment by bolkedebruin** [comment]
@Fokko can you please add unit tests?

@jhsenjaliya some standard connection settings could also apply and be set by an admin (such as kerberos details, master etc). Also unit testing is a bit easier with a wrapper like this (we have more of those).

**Comment by Fokko** [comment]
@bolkedebruin I've added unit tests for both the hook and operator. Please let me know if it sufficient.

**Comment by bolkedebruin** [comment]
@Fokko can please update your pr and rebase?

**Comment by Fokko** [comment]
Oops, I've accidentally rebased against wrong master :-)

I'll update the PR in the beginning of next week. I will then implement:
- Proper killing of yarn jobs by overriding the `on_kill`.
- Adding the `deploy-mode` argument (also to the connection string) as this has become the default since Spark 2.0.

**Comment by Fokko** [comment]
Hi @bolkedebruin 

I've added a couple of things:
- Proper killing of yarn jobs by overriding the on_kill. The output of the jobs will be parsed and the application_id is extracted from it. This only works when logging is set at `INFO` or lower, otherwise the concise and the application_id will not be emitted.
- Adding the deploy-mode argument (also to the connection string) as this has become the default since Spark 2.0.

I've added additional tests and refactored the code a bit. Let me know if it this is sufficient.

For a next pr I would like to add an abstract class for both the spark-{submit, sql} class, and also refactor the spark-sql operator to work with the Airflow connections.

**Comment by bolkedebruin** [comment]
Nice. Lgtm.

**Pull request closed without merging**